### DISCUSSION TITLE
Soroban CLI and the dreaded --arg

### DISCUSSION THREAD

**Discussion by waldmatias** [stellar/sorobanathon#58] [General]
Hi everyone!

I'm not sure if this even qualifies for Sorobanathon, but I think at least it surely is something worth posting here so anyone can take a look and provide feedback!

As part of the Stellar Quest series (geez we're already halfway through) we 'noobs' got to play around with the new soroban-cli. Now, from what I saw, a lot of people were having problems figuring out how to pass function arguments when invoking a contract:
```bash
soroban invoke (...) --fn myFunction --arg nowHowDoIformatWhatImSupposedToSend!!
```
Basically, the cli just gave a very general `value is not parseable to type` output error if you made a mistake. It is good that at least we got some hint that this was a parsing error, but when you add the Built-In types (`Symbol`, `Byte`, `BytesN` and so on) that _could_ be passed, well, you're kind of figuring out how to format them appropriately from the command line. 

So, I guess it made sense that the error message at least also gave a hint of what the _target_ type (for the function's argument) was supposed to be... but yep, this meant deep diving into the code. Honestly, I've figured out that if you want to __really learn__ something, well, you just __have__ to deep-dive (into code); OSS just makes it possible... and that O'Reilly subscription I got awhile back to learn Rust :) 

Anyway, I digress... so back to topic. 

Since the Soroban CLI is on [Stellar's github](https://github.com/stellar/soroban-tools/tree/main/cmd/soroban-cli), it was relatively easy (_yeah, right_) to kind of figure out how the `--arg` feature was implemented. There were a couple of ways to make the changes I wanted, but after doing it the easy way, then the hard-convoluted way, I honestly think I figured out the `Idiomatic Rust`-way of doing it... 
```rust
 Option<> // to the rescue!!
 ```
Basically, I added an `Option<ScSpecType>` to the `InvalidValue` that tells what the arg is supposed to be. Since `ScSpecType` implements a `name()` function that gave a human-readable string, the type name can be included in the error's output.  

If you want to take a look, here's my PR: 
https://github.com/waldmatias/soroban-tools/commit/e6a6238bc0a1d258e26f7c7c7b6209bd432a59db

Not sure if this is going to end up being merged, but at least I had fun!!!

Thanks and I appreciate any/all feedback!
MW

**Comment by achallagundla** [comment]
Heya @waldmatias  ,
Congratulations! Your contribution to Sorobanathon: First Light has been accepted! 🎉
Your submission may qualify for a reward of 800 XLM! One last step: to receive that reward, please fill out this [form](https://stellarform.typeform.com/Sorobanathon) completely and accurately.

Please let us know if you have any questions. (You can also message me on discord @ anuxhya)

All the best,
Anuhya

**Reply by waldmatias** [reply]
Hi @achallagundla ! this is amazing, **thank you!** ref.: form filling will do!
Stay tuned for additional submissions though I feel time is running out!
